### DISCUSSION TITLE
Determining LED on and off timestamps/durations

### DISCUSSION THREAD

**Discussion by cYnd** [gpiozero/gpiozero#1041] [Q&A]
Hello there!

I want to determine when my LED was switched on or off. Therefore i simply want to log to a CSV file when a led was switched on and when it was switched off.
When using the toggle() method, this is quite simple, however I am also using blink on some occasions which makes it not quite simple.

First I thought I could use the 'when_activated' but given the following code:

```
led = gpiozero.LED(gpioPin, active_high = False)
led.when_activated=testPrint

def testPrint(self):
   print("TEST")
```

I will always get this error: 

`AttributeError: 'LED' object has no attribute 'when_activated'`

Which makes sense when I look at [this diagram](https://gpiozero.readthedocs.io/en/stable/_images/device_hierarchy.svg) realising that this method is only available for DigitalInputDevice and not for DigitalOutputDevice.

So is there a way to get the 'when_activated' and 'when_deactivated' methods to a DigitalOutputDevice? Or am I missing something which allows me to track the on times of my LED?
Are there any best practises for this or do I need to cook my own solution for this?

**Comment by lurch** [comment]
Could you explain _why_ you want to log when your LED is switched on or off? That might help us suggest a more suitable solution.

If you want to use `when_activated` I guess you could always go for the "hardware solution" and add a jumper wire between your LED and another GPIO pin that you've set up us a `DigitalInputDevice` :wink:

**Reply by cYnd** [reply]
Well, actually I am not switching a LED rather than a relay card. To this relays are actual lights (230V) connected and I want to calculate how much power is drawn based on the times when the light is on.

I could go for more hardware but since there is already a lot of other stuff going on, I prefer to do it with software :-)

**Reply by lurch** [reply]
Perhaps you just need to create a sub-class of GpioZero's `LED` class? e.g. something like
```python
class LoggedLED(LED):
    def on(self):
        print("Switched on")
        super.on()

    def off(self):
        print("Switched off")
        super.off()
```
(that code above probably won't work, but that should give you the idea)

In fact, if you override the `value` property-setter (rather than the `on()` and `off()` methods), then that _might_ also work for the `blink()` method?

**Reply by cYnd** [reply]
My problem with the `blink()` method is, that I may want to add an switch to turn a blinking LED/Light off. I am using the `blink()` method like a timer, e.g. `blink(120, 0, 1, True)` to let the lights on for 120 seconds. But with the switch I may wnt to turn it off earlier.
Therefore it would have been really handy to have the `when_activated` or `when_deactivated` methods available on the output devices.

My current approach for that is to use a timer which is set to the same time and before logging it checks if the light is already switched off, if not, it reports it.
